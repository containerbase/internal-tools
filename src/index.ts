// Node v16 is missing structuredClone, so we polyfill it here.
import 'core-js/actual/structured-clone';
import run from './runner';

run().catch(console.error);
