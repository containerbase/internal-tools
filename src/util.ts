import { exec as _exec } from '@actions/exec';
import { ExecOptions as _ExecOptions } from '@actions/exec/lib/interfaces';
import log from './utils/logger';
import { getInput, startGroup, endGroup } from '@actions/core';
import { join } from 'path';
import fs from 'fs';
import { promisify } from 'util';
import findUp = require('find-up');

const readFileAsync = promisify(fs.readFile);

export type ExecOptions = _ExecOptions;

export type ExecResult = {
  readonly code: number;
  readonly stdout: string;
  readonly stderr: string;
};

export async function exec(
  cmd: string,
  args: string[],
  options?: ExecOptions
): Promise<ExecResult> {
  let stdout = '';
  let stderr = '';
  let code: number;

  try {
    startGroup(`${cmd} ${args.join(' ')}`);
    code = await _exec(cmd, args, {
      ...options,
      listeners: {
        stdout: (data: Buffer) => {
          stdout += data.toString();
        },
        stderr: (data: Buffer) => {
          stderr += data.toString();
        },
      },
    });
  } finally {
    endGroup();
  }
  if (code) {
    log.error({ code });
    throw new Error('Command failed');
  }

  return { code, stdout, stderr };
}

export function isDryRun(): boolean {
  const val = getInput('dry-run');
  return !!val && val === 'true';
}

/**
 * Get environment variable or empty string.
 * Used for easy mocking.
 * @param key variable name
 */
export function getEnv(key: string): string {
  return process.env[key] ?? '';
}

export async function readJson<T = unknown>(file: string): Promise<T> {
  const path = join(getEnv('GITHUB_WORKSPACE'), file);
  const res = await import(path);
  // istanbul ignore next
  return res?.default ?? res;
}

export async function readFile(file: string): Promise<string> {
  const path = join(getEnv('GITHUB_WORKSPACE'), file);
  return await readFileAsync(path, 'utf8');
}

export const MultiArgsSplitRe = /\s*(?:[;,]|$)\s*/;

export function getArg(name: string, opts?: { required?: boolean }): string;
export function getArg(
  name: string,
  opts?: { required?: boolean; multi: true }
): string[];
export function getArg(
  name: string,
  opts?: { required?: boolean; multi?: boolean }
): string | string[];
export function getArg(
  name: string,
  opts?: { required?: boolean; multi?: boolean }
): string | string[] {
  const val = getInput(name, opts);
  return opts?.multi ? val.split(MultiArgsSplitRe).filter(Boolean) : val;
}

let _pkg: Promise<string | undefined>;

/**
 * Resolve path for a file relative to renovate root directory (our package.json)
 * @param file a file to resolve
 */
export async function resolveFile(file: string): Promise<string> {
  if (!_pkg) {
    _pkg = findUp('package.json', { cwd: __dirname, type: 'file' });
  }
  const pkg = await _pkg;
  // istanbul ignore if
  if (!pkg) {
    throw new Error('Missing package.json');
  }
  return join(pkg, '../', file);
}
