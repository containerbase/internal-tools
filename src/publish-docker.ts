import { getInput } from '@actions/core';
import { exec } from './util';
import chalk from 'chalk';
import log from 'fancy-log';

// sha256:d694b03ba0df63ac9b27445e76657d4ed62898d721b997372aab150ee84e07a1
const shaRe = /sha256:([a-f0-9]{64})/;

async function getImageId(image: string): Promise<string> {
  const res = await exec('docker', ['inspect', "--format='{{.Id}}'", image]);

  const [, id] = shaRe.exec(res.stdout) ?? [];

  if (!id) {
    log.error(res);
    throw new Error('Could not find image id');
  }

  return id;
}

export async function run(): Promise<void> {
  const dryRun = !!getInput('dry-run');
  const image = getInput('image');

  if (!image) {
    throw new Error('Missing image');
  }

  log.info(chalk.blue('Processing image:'), chalk.yellow(image));

  log('Fetch new id');
  const newId = await getImageId(image);

  log('Backup new image');
  await exec('docker', ['tag', image, 'tmp']);
  log('Fetch old image');
  await exec('docker', ['pull', image]);

  log('Fetch old id');
  const oldId = await getImageId(image);

  if (oldId === newId) {
    log('Image uptodate, no push nessessary: ', chalk.yellow(oldId));
    return;
  }

  log('Restore new image');
  await exec('docker', ['tag', 'tmp', image]);
  log('Publish new image');
  if (dryRun) {
    log.warn('DRY-RUN: Would push: ', chalk.yellow(image));
  } else {
    await exec('docker', ['push', image]);
  }
  log.info(chalk.blue('Processing image finished: ', newId));
}
