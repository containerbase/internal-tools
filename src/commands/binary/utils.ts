import { getInput } from '@actions/core';
import is from '@sindresorhus/is';
import { getDefaultVersioning } from 'renovate/dist/modules/datasource';
import { getArch, getArg, getDistro, isDryRun, readJson } from '../../util';
import { readDockerConfig } from '../../utils/config';
import {
  DockerPlatform,
  dockerBuildx,
  dockerRun,
} from '../../utils/docker/common';
import log from '../../utils/logger';
import type { BinaryBuilderConfig, ConfigFile } from '../../utils/types';

export async function getConfig(): Promise<BinaryBuilderConfig> {
  const configFile = getInput('config') || 'builder.json';

  const cfg = await readJson<ConfigFile>(configFile);

  if (!is.object(cfg)) {
    throw new Error('missing-config');
  }

  if (!is.string(cfg.image)) {
    cfg.image = getInput('image', { required: true });
  }

  if (!is.string(cfg.buildArg)) {
    cfg.buildArg = cfg.image.toUpperCase() + '_VERSION';
  }

  await readDockerConfig(cfg);

  return {
    ...cfg,
    ignoredVersions: cfg.ignoredVersions ?? [],
    dryRun: isDryRun(),
    lastOnly: getArg('last-only') == 'true',
    buildArgs: getArg('build-args', { multi: true }),
    versioning: cfg.versioning ?? getDefaultVersioning(cfg.datasource),
  } as BinaryBuilderConfig;
}

export async function createBuilderImage(
  ws: string,
  { buildArgs }: BinaryBuilderConfig
): Promise<void> {
  log('Creating builder image');
  const args = [
    'build',
    '--load',
    '-t',
    'builder',
    '--build-arg',
    `DISTRO=${getDistro()}`,
  ];
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    args.push('--platform', DockerPlatform[arch]);
  }
  if (is.nonEmptyArray(buildArgs)) {
    args.push(...buildArgs.map((b) => `--build-arg=${b}`));
  }
  await dockerBuildx(...args, ws);
}

export async function runBuilder(ws: string, version: string): Promise<void> {
  const args = ['--name', 'builder', '--volume', `${ws}/.cache:/cache`];
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    args.push('--platform', DockerPlatform[arch]);
  }
  await dockerRun(...args, 'builder', version);
}
