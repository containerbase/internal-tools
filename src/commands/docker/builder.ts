import { getInput, setFailed } from '@actions/core';
import is from '@sindresorhus/is';
import * as chalk from 'chalk';
import { getDefaultVersioning } from 'renovate/dist/modules/datasource';
import { get as getVersioning } from 'renovate/dist/modules/versioning';
import { exec, getArg, isDryRun, readJson } from '../../util';
import { BuildsResult, addHostRule, getBuildList } from '../../utils/builds';
import { readDockerConfig } from '../../utils/config';
import { build, publish } from '../../utils/docker';
import { init } from '../../utils/docker/buildx';
import { dockerDf, dockerPrune, dockerTag } from '../../utils/docker/common';
import log from '../../utils/logger';
import type { Config, ConfigFile } from '../../utils/types';

export const MultiArgsSplitRe = /\s*(?:[;,]|$)\s*/;

function createTag(tagSuffix: string | undefined, version: string): string {
  return is.nonEmptyString(tagSuffix) && tagSuffix !== 'latest'
    ? `${version}-${tagSuffix}`
    : version;
}

async function buildAndPush(
  {
    imagePrefix,
    image,
    buildArg,
    buildArgs,
    buildOnly,
    cache,
    dryRun,
    tagSuffix,
    versioning,
    majorMinor,
    prune,
  }: Config,
  tobuild: BuildsResult
): Promise<void> {
  const builds: string[] = [];
  const failed: string[] = [];
  const ver = getVersioning(versioning);
  const versionsMap = new Map<string, string>();
  if (majorMinor) {
    for (const version of tobuild.versions) {
      const minor = ver.getMinor(version);
      const major = ver.getMajor(version);
      const isStable = ver.isStable(version);

      if (isStable && is.number(major) && `${major}` !== version) {
        versionsMap.set(`${major}`, version);
      }

      if (
        isStable &&
        is.number(major) &&
        is.number(minor) &&
        `${major}.${minor}` !== version
      ) {
        versionsMap.set(`${major}.${minor}`, version);
      }
    }
  }

  await exec('df', ['-h']);

  for (const version of tobuild.versions) {
    const tag = createTag(tagSuffix, version.replace(/\+.+/, ''));
    const imageVersion = `${imagePrefix}/${image}:${tag}`;
    log(`Building ${imageVersion}`);
    try {
      const minor = ver.getMinor(version);
      const major = ver.getMajor(version);
      const cacheTags: string[] = [tagSuffix ?? 'latest'];
      const tags: string[] = [];

      if (
        is.number(major) &&
        majorMinor &&
        versionsMap.get(`${major}`) === version
      ) {
        const nTag = createTag(tagSuffix, `${major}`);
        cacheTags.push(nTag);
        tags.push(nTag);
      }

      if (
        is.number(major) &&
        is.number(minor) &&
        majorMinor &&
        versionsMap.get(`${major}.${minor}`) === version
      ) {
        const nTag = createTag(tagSuffix, `${major}.${minor}`);
        cacheTags.push(nTag);
        tags.push(nTag);
      }

      if (version === tobuild.latestStable) {
        tags.push(tagSuffix ?? 'latest');
      }

      await build({
        image,
        imagePrefix,
        tag,
        cache,
        cacheTags,
        buildArgs: [...(buildArgs ?? []), `${buildArg}=${version}`],
        dryRun,
      });

      if (!buildOnly) {
        await publish({ image, imagePrefix, tag, dryRun });
        const source = tag;

        for (const tag of tags) {
          log(`Publish ${source} as ${tag}`);
          await dockerTag({ image, imagePrefix, src: source, tgt: tag });
          await publish({ image, imagePrefix, tag, dryRun });
        }
      }

      log(`Build ${imageVersion}`);
      builds.push(version);
    } catch (err) {
      log.error(err);
      failed.push(version);
    }

    await dockerDf();
    await exec('df', ['-h']);

    if (prune) {
      await dockerPrune();
      await exec('df', ['-h']);
    }
  }

  if (builds.length) {
    log('Build list:' + builds.join(' '));
  }

  if (failed.length) {
    log.warn('Failed list:' + failed.join(' '));
    throw new Error('failed');
  }
}

async function generateImages(config: Config): Promise<void> {
  const buildList = await getBuildList(config);

  if (!buildList?.versions.length) {
    setFailed(`No versions found.`);
    return;
  }

  await buildAndPush(config, buildList);
}

export async function run(): Promise<void> {
  const dryRun = isDryRun();
  const configFile = getInput('config') || 'builder.json';

  const cfg = await readJson<ConfigFile>(configFile);

  if (!is.object(cfg)) {
    throw new Error('missing-config');
  }

  // TODO: validation
  if (!is.string(cfg.image)) {
    cfg.image = getInput('image', { required: true });
  }

  if (!is.string(cfg.buildArg)) {
    cfg.buildArg = cfg.image.toUpperCase() + '_VERSION';
  }

  await readDockerConfig(cfg);

  const config: Config = {
    ...cfg,
    imagePrefix: getArg('image-prefix')?.replace(/\/$/, '') || 'renovate',
    image: cfg.image,
    depName: cfg.depName ?? cfg.image,
    buildArg: cfg.buildArg,
    buildArgs: getArg('build-args', { multi: true }),
    tagSuffix: getArg('tag-suffix') || undefined,
    ignoredVersions: cfg.ignoredVersions ?? [],
    dryRun,
    lastOnly: getInput('last-only') == 'true',
    buildOnly: getInput('build-only') == 'true',
    majorMinor: getArg('major-minor') !== 'false',
    prune: getArg('prune') === 'true',
    versioning: cfg.versioning ?? getDefaultVersioning(cfg.datasource),
  };

  if (dryRun) {
    log('GitHub Actions branch detected - Force building latest, no push');
    config.lastOnly = true;
  }
  log('config:', JSON.stringify(config));

  const token = getArg('token');

  if (token) {
    addHostRule({ matchHost: 'github.com', token });
  }

  await init();

  await generateImages(config);

  log.info(chalk.blue('Processing done:', config.image));
}
