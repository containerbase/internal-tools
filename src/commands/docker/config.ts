import { init } from '../../utils/docker/buildx';

export async function run(): Promise<void> {
  await init();
}
