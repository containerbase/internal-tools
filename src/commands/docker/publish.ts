import { getInput } from '@actions/core';
import { isDryRun } from '../../util';
import { publish } from '../../utils/docker';
import log from '../../utils/logger';

export const MultiArgsSplitRe = /\s*(?:;|$)\s*/;
export async function run(): Promise<void> {
  const dryRun = isDryRun();
  let image = getInput('image', { required: true });
  let tags = getInput('tags')?.split(MultiArgsSplitRe).filter(Boolean);

  if (!tags?.length) {
    tags = ['latest'];
  }

  if (!image) {
    throw new Error('Missing image');
  }

  // istanbul ignore if
  if (image.startsWith('renovate/')) {
    log.warn('Image prefix can be removed:', image);
    image = image.substring(9);
  }

  for (const tag of tags) {
    await publish({ image, tag, dryRun });
  }
}
