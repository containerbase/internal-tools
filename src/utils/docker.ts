import is from '@sindresorhus/is';
import * as chalk from 'chalk';
import * as delay from 'delay';
import got, { HTTPError, Headers } from 'got';
import * as wwwAuthenticate from 'www-authenticate';
import { exists } from '../util';
import { cosign } from './cosign/common';
import { docker } from './docker/common';
import log from './logger';
import { ExecError } from './types';

export function splitImage(image: string): {
  registry: string;
  repository: string;
} {
  const parts = image.split('/');
  if (parts.length > 2) {
    return { registry: parts[0], repository: parts.slice(1).join('/') };
  } else {
    return { registry: 'index.docker.io', repository: image };
  }
}

export type DockerManifestV2 = {
  schemaVersion: number;
  config: {
    digest: string;
  };
};

export async function getAuthHeaders(
  registry: string,
  repository: string
): Promise<Headers> {
  try {
    const apiCheckUrl = `https://${registry}/v2/`;
    const apiCheckResponse = await got(apiCheckUrl, { throwHttpErrors: false });
    if (apiCheckResponse.headers['www-authenticate'] === undefined) {
      return {};
    }
    const authenticateHeader = new wwwAuthenticate.parsers.WWW_Authenticate(
      apiCheckResponse.headers['www-authenticate']
    );

    const authUrl = `${authenticateHeader.parms.realm}?service=${authenticateHeader.parms.service}&scope=repository:${repository}:pull`;
    const authResponse = (
      await got<{ token?: string; access_token?: string }>(authUrl, {
        responseType: 'json',
      })
    ).body;

    const token = authResponse.token || authResponse.access_token;
    if (!token) {
      throw new Error('Failed to obtain docker registry token');
    }
    return {
      authorization: `Bearer ${token}`,
    };
  } catch (err) {
    log.error(chalk.red('auth error'), (err as Error).message);
    throw new Error('Failed to obtain docker registry token');
  }
}

export enum DockerContentType {
  ManifestV1 = 'application/vnd.docker.distribution.manifest.v1+json',
  ManifestV1Signed = 'application/vnd.docker.distribution.manifest.v1+prettyjws',
  ManifestV2 = 'application/vnd.docker.distribution.manifest.v2+json',
}

const shaRe = /(sha256:[a-f0-9]{64})/;

export async function getRemoteImageId(
  registry: string,
  repository: string,
  tag = 'latest'
): Promise<string> {
  const headers = await getAuthHeaders(registry, repository);
  headers.accept = DockerContentType.ManifestV2;
  const url = `https://${registry}/v2/${repository}/manifests/${tag}`;

  try {
    const resp = await got<DockerManifestV2>(url, {
      headers,
      responseType: 'json',
    });

    switch (resp.headers['content-type']) {
      case DockerContentType.ManifestV2:
        return resp.body.config.digest;

      case DockerContentType.ManifestV1:
      case DockerContentType.ManifestV1Signed:
        // something wrong, we need to overwrite existing
        log.warn(
          chalk.yellow('Wrong response'),
          `Wrong response: ${resp.headers['content-type'] as string}`
        );
        return '<error>';

      default:
        throw new Error(
          `Unsupported response: ${resp.headers['content-type'] as string}`
        );
    }
  } catch (e) {
    if (e instanceof HTTPError && e.response.statusCode === 404) {
      // no image published yet
      return '<none>';
    }
    log.error(chalk.red('request error'), (e as Error).message);
    throw new Error('Could not find remote image id');
  }
}

export async function getLocalImageId(
  image: string,
  tag = 'latest'
): Promise<string> {
  const res = await docker('inspect', "--format='{{.Id}}'", `${image}:${tag}`);

  const [, id] = shaRe.exec(res.stdout) ?? [];

  if (!id) {
    log.error(res);
    throw new Error('Could not find local image id');
  }

  return id;
}

export type BuildOptions = {
  image: string;
  imagePrefix: string;
  cache?: string;
  cacheTags?: string[];
  tag?: string;
  dryRun?: boolean;
  buildArgs?: string[];
};

const errors = [
  'unexpected status: 400 Bad Request',
  ': no response',
  'error writing layer blob',
];

function canRetry(err: ExecError): boolean {
  return errors.some((str) => err.stderr.includes(str));
}

export async function build({
  image,
  imagePrefix,
  cache,
  cacheTags,
  tag = 'latest',
  dryRun,
  buildArgs,
}: BuildOptions): Promise<void> {
  const args = [
    'buildx',
    'build',
    '--load',
    `--tag=${imagePrefix}/${image}:${tag}`,
  ];

  if (is.nonEmptyArray(buildArgs)) {
    args.push(...buildArgs.map((b) => `--build-arg=${b}`));
  }

  if (is.string(cache)) {
    const cacheImage = `${imagePrefix}/${cache}:${image.replace(/\//g, '-')}`;
    args.push(`--cache-from=${cacheImage}-${tag}`);

    if (is.nonEmptyArray(cacheTags)) {
      for (const ctag of cacheTags) {
        args.push(`--cache-from=${cacheImage}-${ctag}`);
      }
    }

    if (!dryRun) {
      args.push(`--cache-to=type=registry,ref=${cacheImage}-${tag},mode=max`);
    }
  }

  for (let build = 0; ; build++) {
    try {
      await docker(...args, '.');
      break;
    } catch (e) {
      if (e instanceof ExecError && canRetry(e) && build < 2) {
        log.error(chalk.red(`docker build error on try ${build}`), e);
        await delay(5000);
        continue;
      }
      throw e;
    }
  }
}

type PublishOptions = {
  image: string;
  imagePrefix: string;
  tag: string;
  dryRun?: boolean;
};

export async function publish({
  image,
  imagePrefix,
  tag,
  dryRun,
}: PublishOptions): Promise<void> {
  const imageName = `${imagePrefix}/${image}`;
  const fullName = `${imageName}:${tag}`;
  log.info(chalk.blue('Processing image:'), chalk.yellow(fullName));

  log('Fetch new id');
  const newId = await getLocalImageId(imageName, tag);

  log('Fetch old id');
  const { registry, repository } = splitImage(imageName);
  const oldId = await getRemoteImageId(registry, repository, tag);

  if (oldId === newId) {
    log('Image uptodate, no push nessessary:', chalk.yellow(oldId));
    return;
  }

  log('Publish new image', `${oldId} => ${newId}`);
  if (dryRun) {
    log.warn(chalk.yellow('[DRY_RUN]'), chalk.blue('Would push:'), fullName);
  } else {
    await docker('push', fullName);
    await sign(fullName);
  }
  log.info(chalk.blue('Processing image finished:', newId));
}

async function sign(fullName: string): Promise<void> {
  if (!(await exists('cosign'))) {
    log.warn('Cosign is not installed. Skipping container signing');
    return;
  }

  log('Signing image', fullName);
  await cosign('sign', fullName);
}
