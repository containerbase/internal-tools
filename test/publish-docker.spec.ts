import * as _core from '@actions/core';
import { getName, mocked } from './utils';
import * as _docker from '../src/utils/docker';
import * as _utils from '../src/util';
import { run } from '../src/publish-docker';

jest.mock('@actions/core');
jest.mock('@actions/github');
jest.mock('fancy-log');
jest.mock('../src/util');
jest.mock('../src/utils/docker');

const core = mocked(_core);
const utils = mocked(_utils);
const docker = mocked(_docker);
const image = 'renovate/base';
const digest =
  'sha256:d694b03ba0df63ac9b27445e76657d4ed62898d721b997372aab150ee84e07a1';

describe(getName(__filename), () => {
  beforeEach(() => {
    jest.clearAllMocks();
    core.getInput.mockReturnValueOnce('').mockReturnValueOnce(image);
    docker.getLocalImageId.mockResolvedValueOnce(digest);
  });

  it('uptodate', async () => {
    docker.getRemoteImageId.mockResolvedValueOnce(digest);

    await run();

    expect(docker.getLocalImageId).toHaveBeenCalledWith(image);
    expect(docker.getRemoteImageId).toHaveBeenCalledWith(image);
    expect(utils.exec).not.toHaveBeenCalled();
  });

  it('updates image', async () => {
    docker.getLocalImageId.mockResolvedValueOnce(
      'sha256:d694b03ba0df63ac9b27445e76657d4ed62898d721b997372aab150ee84e07a2'
    );

    await run();

    expect(docker.getLocalImageId).toHaveBeenCalledWith(image);
    expect(docker.getRemoteImageId).toHaveBeenCalledWith(image);
    expect(utils.exec).toBeCalledTimes(1);
  });

  it('updates image (dry-run)', async () => {
    core.getInput.mockReset();
    core.getInput.mockReturnValueOnce('true').mockReturnValueOnce(image);
    docker.getLocalImageId.mockResolvedValueOnce(
      'sha256:d694b03ba0df63ac9b27445e76657d4ed62898d721b997372aab150ee84e07a2'
    );
    await run();

    expect(docker.getLocalImageId).toHaveBeenCalledWith(image);
    expect(docker.getRemoteImageId).toHaveBeenCalledWith(image);
    expect(utils.exec).not.toHaveBeenCalled();
  });

  it('throws no image', async () => {
    core.getInput.mockReset();
    await expect(run()).rejects.toThrow('Missing image');
  });
});
