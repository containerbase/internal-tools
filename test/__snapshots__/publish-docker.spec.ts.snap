// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`missing image 1`] = `[Error: Missing image]`;

exports[`no image id 1`] = `[Error: Could not find image id]`;
